packages/flutter_tools/lib/src/build_system/targets/web.dart

```diff
@@ -270,29 +279,91 @@ class Dart2JSTarget extends Target {
   }
 }
 
-/// Unpacks the dart2js compilation and resources to a given output directory.
+class Dart2WasmTarget extends Dart2WebTarget {
+  Dart2WasmTarget(super.webRenderer);
+
+  @override
+  Future<void> build(Environment environment) async {
+    final String? buildModeEnvironment = environment.defines[kBuildMode];
+    if (buildModeEnvironment == null) {
+      throw MissingDefineException(kBuildMode, name);
+    }
+    final BuildMode buildMode = getBuildModeForName(buildModeEnvironment);
+    final Artifacts artifacts = globals.artifacts!;
+    final File outputWasmFile = environment.buildDir.childFile('main.dart.wasm');
+    final String dartSdkPath = artifacts.getArtifactPath(Artifact.engineDartSdkPath, platform: TargetPlatform.web_javascript);
+    final String dartSdkRoot = environment.fileSystem.directory(dartSdkPath).parent.uri.toString();
+
+    final List<String> compilationArgs = <String>[
+      artifacts.getArtifactPath(Artifact.engineDartAotRuntime, platform: TargetPlatform.web_javascript),
+      '--disable-dart-dev',
+      artifacts.getArtifactPath(Artifact.dart2wasmSnapshot, platform: TargetPlatform.web_javascript),
+      if (buildMode == BuildMode.profile)
+        '-Ddart.vm.profile=true'
+      else
+        '-Ddart.vm.product=true',
+      ...decodeCommaSeparated(environment.defines, kExtraFrontEndOptions),
+      for (final String dartDefine in decodeDartDefines(environment.defines, kDartDefines))
+        '-D$dartDefine',
+      '--packages=.dart_tool/package_config.json',
+      '--dart-sdk=$dartSdkPath',
+      '--multi-root-scheme',
+      'org-dartlang-sdk',
+      '--multi-root',
+      artifacts.getHostArtifact(HostArtifact.flutterWebSdk).uri.toString(),
+      '--multi-root',
+      dartSdkRoot,
+      '--libraries-spec',
+      artifacts.getHostArtifact(HostArtifact.flutterWebLibrariesJson).path,
+
+      environment.buildDir.childFile('main.dart').path, // dartfile
+      outputWasmFile.path,
+    ];
+    final ProcessResult compileResult = await globals.processManager.run(compilationArgs);
+    if (compileResult.exitCode != 0) {
+      throw Exception(_collectOutput(compileResult));
+    }
+  }
+
+  @override
+  Source get compilerSnapshot => const Source.artifact(Artifact.dart2wasmSnapshot);
+
+  @override
+  String get name => 'dart2wasm';
+
+  @override
+  List<Source> get outputs => const <Source>[
+    Source.pattern('{OUTPUT_DIR}/main.dart.wasm'),
+  ];
+}
+
+/// Unpacks the dart2js or dart2wasm compilation and resources to a given
+/// output directory.
 class WebReleaseBundle extends Target {
-  const WebReleaseBundle(this.webRenderer);
+  const WebReleaseBundle(this.webRenderer, this.isWasm);
 
   final WebRendererMode webRenderer;
+  final bool isWasm;
+
+  String get outputFileName => isWasm ? 'main.dart.wasm' : 'main.dart.js';
```

Comment:
I think with `--wasm` we always want to compile to both because for a while apps will need to be able to compile and run in JS mode in browsers that don't have WasmGC yet.

Reply:
Alternatively, instead of `--wasm` we could have `--targets=wasm,js` for devs to choose from (e.g. those fully committed to wasm may have code that only works in wasm and vice versa). `--targets=js` would be the default.

Reply:
Two things: 

* I think it's important to be able to build wasm on its own, if not just for my own sanity. I don't want to build the dart2js target over and over when I'm iterating on the wasm side.
* A dart2js build and a dart2wasm build can't actually output to the same directory (at least right now), because dart2wasm actually generates a `main.dart.js` that bootstraps the wasm module. We can adjust this later if we want to do some kind of auto-switcheroo depending on detecting the browser's WasmGC support, but it's not really a priority.

For these reasons... I'd like to avoid mixing the dart2js and dart2wasm builds for the time being. I think we can adjust this later as this becomes a little less bleeding-edge.

Reply:
Yeah that's a good idea. I'll do that.

Reply:
I don't mean to be prescriptive about what exact command-line options we have. My main request is that we don't paint ourselves into a corner using an option/target design that we'll have to deprecate immediately as soon as we get close to releasing this in stable. To that end I think it's useful to think about scenarios we want to support and make sure the command-line option structure we choose _will_ support it (even if it won't support everything immediately). Here are the scenarios I'm thinking of:

* **The default**: we output both JS and wasm and at runtime pick one, depending on browser capabilities.
* **Wasm only**: the developer doesn't care about the JS option; perhaps their code relies on wasm-only features, such as numerics and SIMD, not available in JS.
* **JS only**: the developer is not yet ready to adopt wasm (e.g. their plugins are not ready) and wants JS only, even in browsers that do support WasmGC.

I'm only raising awareness. If the current approach supports these modes, it's fine with me.

Reply:
I agree with all your points. This is just the very first pass at the command line stuff, I'll keep this stuff in mind in subsequent revisions.